rewrite to lexer/parser approach

hairmare opened the pull request:
This rewrites most of the parser to a doctrine/lexer based approach.

This should be able to pave the way for #6 as well as contribute to GRV-1289.

Implements a heap of rql operations in their basic non sugary form.
- [x] `eq()`, `ne()`, `lt()`, `gt()`, `lte()` and `gte()` support
- [x] `and()` and `or()` support
- [x] implement `like()` support
- [x] fix failing `or()` test
- [x] implement `in()`, `out()`
- [x] implement `limit()`

These are things we also need to do. I'd rather do them in different PRs so we can get 2.0.0-alpha1 out of the door soonish.
- refactor AST -> we need `AST\EqOperation` based on `AST\AbstractPropertyOperation` and so on. This also adds some more interfaces. The types we need are probably propertyOp, queryOp, visitOp and sortOp, though the last two need checking with other possible methods we'll be needing. (would be better suited for a separate PR)
- implement syntax sugar variants, probably needs some parser refactoring
- re-integrate with graviton by way of a wrapper-bundle (ie. like serialiser does, also tracked as GRV-1283, [http://gnugat.github.io/2014/10/29/sf2-bundle-standalone.html]) 
- fork mongo visitor into libgraviton/php-rql-parser-visitor-mongo-odm so everyone is treated the same

This also needs GRV-1283 to be useable with graviton. Along with GRV-1283 it fixes GRV-1289.

Commit:
partial lexer implementation

Commit:
initial work on lexer based parser

Commit:
whitespace in T_STRING

Commit:
implement ne() and refactor some

Commit:
add AND support to lexer

Commit:
basic AND support in parser

Commit:
strip out old tests for now

Commit:
rename lib dir to src dir

just so this is like in all the other projects.

Commit:
scaffolding and testing for mongo visitor

Commit:
allow integers in parser

Commit:
fix indent

Commit:
implement some searches

Commit:
support or() in lexer

Commit:
parse or()

Commit:
lt(),gt(),lte(),gte() support in lexer/parser

Commit:
initial work on and/or expressions

Commit:
add new operators to visitor

Commit:
sort tokens

Commit:
sort() in parser

Commit:
partial fix to sort() not working

Commit:
refactor parser and visitor

Commit:
fix travis and PSR issues

Commit:
revamp README

I plan on re-adding some examples as soon as the API stabilzes again.

Commit:
bump version to 2.0.0-dev

Commit:
remove old Queriable style API

Commit:
fix typo

Commit:
some refactoring

Commit:
more refactoring

Commit:
fix parsing for multi sort

Commit:
refactor lexer

Commit:
more refactoring

Commit:
remove version as per composer recommendations

Commit:
support like() in lexer and parser

Commit:
implement like() for MongoODM visitor

Commit:
fix or() test

Commit:
support limit in lexer and parser

Commit:
implement limit() for mongodb

Commit:
Revert "fix or() test"

This reverts commit 2639a3bd7ce7ba3eeaa84d73cbd44e8a99d6764e.

narcoticfresh commented on src/Graviton/Rql/Visitor/MongoOdm.php:
```diff
@@ -0,0 +1,121 @@
+        'gt' => 'gt',
+        'lte' => 'lte',
+        'gte' => 'gte'
+);
```
The keys here should reference to the corresponding Lexer Constants..

hairmare replied:
These do not use the Lexer::T_\* constants as not to couple the visitor to the Lexer. I think the visitor should only depend on the AST. The "magic" strings here come from the ASTs `getType()` operation.

I should probably refactor the AST so it can express this as a composition of various classes. That would then lead to `AST\Operation\Eq` and so on. I'm not convinced that refactoring the AST will make the code more reusable though.

narcoticfresh commented on src/Graviton/Rql/Visitor/MongoOdm.php:
```diff
@@ -0,0 +1,121 @@
+        'sort' => 'visitSort',
+        'like' => 'visitLike',
+        'limit' => 'visitLimit',
+    );
```
Same here, keys should refer to some constants..

hairmare replied:
Same as above, the Visitor should not be coupled to the Lexer. This is based on the AST.
I'll be refactoring the AST as part of a different PR.

narcoticfresh commented on src/Graviton/Rql/Visitor/MongoOdm.php:
```diff
@@ -0,0 +1,121 @@
+            $expr = $expr->equals($operation->getValue());
+        } elseif ($operation->getName() == 'ne') {
+            $expr = $expr->notEqual($operation->getValue());
+        }
```
'eq' and 'ne' and so on should also refer to constants..

hairmare replied:
Again, this should not couple the visitor to the lexer. This method does need work to be usable though.

narcoticfresh commented:
What about defining constants in the interface which the Visitor can refer to?
Then we had a defined set of operations and types the Visitors can use and it's not coupled to the Lexer.. even if it's duplication it allows for a better style in the Visitor..

Commit:
support in(), out() in lexer

Commit:
implement in() and out() in parser

Commit:
implement in() and out() searches

hairmare commented:
I don't know about using `const` in an `interface`. I was under the impression that that would be invalid.

What do you think about refactoring the AST so the info you're looking at is explicitly added to the graph?

Commit:
fix non interface use

Commit:
remove tree traversal from visitor call

for now each visitor should traverse the tree on his own as needed. I might
change this if I get around to refactoring the AST into something using more
classes.

Commit:
E_STRICT fix

seems the better fit here since we need a default skip anyhow

Commit:
mark "eq OR with sugar" test as skipped

I'm not currently implementing sugar and will reactivate as soon as I start implemnting sugar.

Commit:
casting for limit and skip values

hairmare commented:
I just removed the [WIP] stamp (we should switch to WIP labels btw).

This PR is getting larger as it should already ;)

To summarize:
- The is practically a from the grounds up rewrite. I did keep small parts though.
- Its based on doctrine/lexer that has been stable for a loooong time now and is heavily depended on by most doctrine projects (by way of the annotation-parser btw).
- This needs a refactor of the internal AST asap.
- more stuff thats mentioned in the PRs original text (and either covered by GRV or things I'll create issues on)

Also, this isn't strictly at a usable state right now. We need to write an additional bundle for integrating this with symfony. This only has doctrine/lexer and the querybuilder from doctrine/mongo-odm as runtime dependency. The internal AST refactor does not interact with this.

Basically, I want to merge it soonish so we can release these results as `2.0.0-alpha1` and get on with building a properly integrated 2.0.0 release.

@narcoticfresh I started an AST refactor over here: 66e3fb7c10b6a70551a50273eda73e1b766070bf
We should talk the AST structure through, before I go all in.

Commit:
Merge branch 'develop' into feature/lexer-parser-approach

narcoticfresh commented:
Well, we're mixing too much at once ;-)
First off, a Symfony integration obviously needs to be a separate project with a separate composer.json and all.. not scope of this project here..

Regarding the AST refactoring: I don't think we should blow this up too much. Wouldn't it be more leaner to just define an abstract class somewhere that just contains constants to which both parts can refer?

hairmare commented:
The symfony bundle won't be part of this PR. I'm just mentioning it since we need it so php-rql-parser is only with graviton again.

I've identified some grouping that make sense in the AST (making for easier visitors that don't use magic strings). We can discuss those in a new PR as soon as this gets merged.
